docs/docs/deploy/2-configuration.md
# Configuration

## Config schema reference

| Property                 | Environment variable       | Property type                                                           | Default | Description                                                                                        |
|--------------------------|----------------------------|-------------------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------|
| `syncPeriod`             | `SYNC_PERIOD`              | string                                                                  | "5m"    | SyncPeriod determines the minimum frequency at which watched resources are reconciled.             |
| `watchNamespaces`        | `WATCH_NAMESPACES`         | string                                                                  | ""      | A comma-separated list of namespaces to watch for resources. If empty, all namespaces are watched. |
| `customResourceSelector` | `CUSTOM_RESOURCE_SELECTOR` | string                                                                  | ""      | A selector for custom resources.                                                                   |
| `enableHTTP2`            | `ENABLE_HTTP2`             | boolean                                                                 | true    | Enables or disables HTTP/2 support for the service.                                                |
| `metrics`                |                            | [MetricsConfig](#metricsconfig-schema-reference)                        |         | Configuration for metrics collection and exposure.                                                 |
| `probe`                  |                            | [ProbeConfig](#probeconfig-schema-reference)                            |         | Configuration for health and readiness probes.                                                     |
| `pprof`                  |                            | [PprofConfig](#pprofconfig-schema-reference)                            |         | Configuration for pprof profiling endpoints.                                                       |
| `logging`                |                            | [LoggingConfig](#loggingconfig-schema-reference)                        |         | Configuration for logging behavior and format.                                                     |
| `leaderElection`         |                            | [LeaderElectionConfig](#leaderelectionconfig-schema-reference)          |         | Configuration for leader election in a multi-instance setup.                                       |
| `globalController`       |                            | [GlobalControllerConfig](#globalcontrollerconfig-schema-reference)      |         | Global configuration for controllers managing resources.                                           |
| `specificControllers`    |                            | [SpecificControllerConfig](#specificcontrollersconfig-schema-reference) |         | Configuration specific to individual controllers.                                                  |

## MetricsConfig schema reference

Used in [Config](#config-schema-reference)

| Property  | Environment variable | Property type | Default | Description                                          |
|-----------|----------------------|---------------|---------|------------------------------------------------------|
| `enabled` | `METRICS_ENABLED`    | boolean       | false   | Enables or disables the metrics server.              |
| `port`    | `METRICS_PORT`       | integer       | 8080    | The port on which the metrics server listens.        |
| `secure`  | `METRICS_SECURE`     | boolean       | false   | Enables or disables secure (HTTPS) metrics endpoint. |

## ProbeConfig schema reference

Used in [Config](#config-schema-reference)

| Property                | Environment variable            | Property type | Default   | Description                                        |
|-------------------------|---------------------------------|---------------|-----------|----------------------------------------------------|
| `enabled`               | `PROBE_ENABLED`                 | boolean       | false     | Enables or disables health and readiness probes.   |
| `port`                  | `PROBE_PORT`                    | integer       | 8082      | The port on which the probe endpoints are exposed. |
| `readinessEndpointName` | `PROBE_READINESS_ENDPOINT_NAME` | string        | "readyz"  | The endpoint path for readiness checks.            |
| `livenessEndpointName`  | `PROBE_LIVENESS_ENDPOINT_NAME`  | string        | "healthz" | The endpoint path for liveness checks.             |

## PprofConfig schema reference

Used in [Config](#config-schema-reference)

| Property  | Environment variable | Property type | Default | Description                                        |
|-----------|----------------------|---------------|---------|----------------------------------------------------|
| `enabled` | `PPROF_ENABLED`      | boolean       | false   | Enables or disables the pprof profiling endpoints. |
| `port`    | `PPROF_PORT`         | integer       | 8081    | The port on which the pprof endpoints are exposed. |

## LoggingConfig schema reference

Used in [Config](#config-schema-reference)

| Property | Environment variable | Property type                                     | Default | Description                                                                       |
|----------|----------------------|---------------------------------------------------|---------|-----------------------------------------------------------------------------------|
| `level`  | `LOG_LEVEL`          | string (one of [debug, info, warn, error, fatal]) | "info"  | The log level for the application. Determines the verbosity of logs.              |
| `format` | `LOG_FORMAT`         | string (one of [console, json])                   | "json"  | The format of the logs. Can be human-readable (`console`) or structured (`json`). |

## LeaderElectionConfig schema reference

Used in [Config](#config-schema-reference)

| Property        | Environment variable             | Property type | Default                            | Description                                                 |
|-----------------|----------------------------------|---------------|------------------------------------|-------------------------------------------------------------|
| `enabled`       | `LEADER_ELECTION_ENABLED`        | boolean       | false                              | Enables or disables leader election.                        |
| `id`            | `LEADER_ELECTION_ID`             | string        | "casdoor-operator-leader-election" | The unique identifier for the leader election process.      |
| `namespace`     | `LEADER_ELECTION_NAMESPACE`      | string        | "default"                          | The namespace where leader election resources are created.  |
| `retryPeriod`   | `LEADER_ELECTION_RETRY_PERIOD`   | string        | "2s"                               | The interval at which leader election attempts are retried. |
| `renewDeadline` | `LEADER_ELECTION_RENEW_DEADLINE` | string        | "10s"                              | The deadline by which the leader must renew its lease.      |
| `leaseDuration` | `LEADER_ELECTION_LEASE_DURATION` | string        | "15s"                              | The duration for which a leader's lease is valid.           |

## GlobalControllerConfig schema reference

Used in [Config](#config-schema-reference)

| Property                  | Environment variable                          | Property type  | Default | Description                                                        |
|---------------------------|-----------------------------------------------|----------------|---------|--------------------------------------------------------------------|
| `skipNameValidation`      | `GLOBAL_CONTROLLER_SKIP_NAME_VALIDATION`      | boolean        | false   | Skips validation of resource names if enabled.                     |
| `groupKindConcurrency`    | `GLOBAL_CONTROLLER_GROUP_KIND_CONCURRENCY`    | map[string]int |         | Specifies the concurrency limits for different resource types.     |
| `maxConcurrentReconciles` | `GLOBAL_CONTROLLER_MAX_CONCURRENT_RECONCILES` | integer        | 1       | The maximum number of concurrent reconciliation operations.        |
| `cacheSyncTimeout`        | `GLOBAL_CONTROLLER_CACHE_SYNC_TIMEOUT`        | string         | "2m"    | The timeout for syncing the cache with the API server.             |
| `needLeaderElection`      | `GLOBAL_CONTROLLER_NEED_LEADER_ELECTION`      | boolean        | true    | Indicates whether leader election is required for this controller. |

## SpecificControllersConfig schema reference

Used in [Config](#config-schema-reference)

| Property  | Environment variable | Property type                                                        | Default | Description                                       |
|-----------|----------------------|----------------------------------------------------------------------|---------|---------------------------------------------------|
| `casdoor` |                      | [CasdoorControllerConfig](#casdoorcontrollerconfig-schema-reference) |         | Configuration specific to the Casdoor controller. |

## CasdoorControllerConfig schema reference

Used in [SpecificControllersConfig](#specificcontrollersconfig-schema-reference)

| Property                  | Environment variable                           | Property type | Default | Description                                                                       |
|---------------------------|------------------------------------------------|---------------|---------|-----------------------------------------------------------------------------------|
| `skipNameValidation`      | `CASDOOR_CONTROLLER_SKIP_NAME_VALIDATION`      | boolean       | false   | Skips validation of Casdoor resource names if enabled.                            |
| `maxConcurrentReconciles` | `CASDOOR_CONTROLLER_MAX_CONCURRENT_RECONCILES` | integer       | 1       | The maximum number of concurrent reconciliation operations for Casdoor resources. |
| `cacheSyncTimeout`        | `CASDOOR_CONTROLLER_CACHE_SYNC_TIMEOUT`        | string        | "2m"    | The timeout for syncing the cache with the API server for Casdoor resources.      |
| `needLeaderElection`      | `CASDOOR_CONTROLLER_NEED_LEADER_ELECTION`      | boolean       | true    | Indicates whether leader election is required for the Casdoor controller.         |


config/config.go
/*
Copyright 2025 ptrvsrg.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//nolint:lll

package config

import "time"

type Config struct {
	SyncPeriod             time.Duration `yaml:"syncPeriod" env:"SYNC_PERIOD" env-default:"5m"`
	WatchNamespaces        string        `yaml:"watchNamespaces" env:"WATCH_NAMESPACES"`
	CustomResourceSelector string        `yaml:"customResourceSelector" env:"CUSTOM_RESOURCE_SELECTOR"`
	EnableHTTP2            *bool         `yaml:"enableHTTP2" env:"ENABLE_HTTP2" env-default:"true"`

	Metrics             MetricsConfig             `yaml:"metrics"`
	Probe               ProbeConfig               `yaml:"probe"`
	Pprof               PprofConfig               `yaml:"pprof"`
	Logging             LoggingConfig             `yaml:"logging"`
	LeaderElection      LeaderElectionConfig      `yaml:"leaderElection"`
	GlobalController    GlobalControllerConfig    `yaml:"globalController"`
	SpecificControllers SpecificControllersConfig `yaml:"specificControllers"`
}

type LoggingConfig struct {
	Level  string `yaml:"level" env:"LOG_LEVEL" env-default:"info"`
	Format string `yaml:"format" env:"LOG_FORMAT" env-default:"json"`
}

type LeaderElectionConfig struct {
	Enabled       *bool         `yaml:"enabled" env:"LEADER_ELECTION_ENABLED" env-default:"false"`
	ID            string        `yaml:"id" env:"LEADER_ELECTION_ID" env-default:"casdoor-operator-leader-election"`
	Namespace     string        `yaml:"namespace" env:"LEADER_ELECTION_NAMESPACE" env-default:"default"`
	RetryPeriod   time.Duration `yaml:"retryPeriod" env:"LEADER_ELECTION_RETRY_PERIOD" env-default:"2s"`
	RenewDeadline time.Duration `yaml:"renewDeadline" env:"LEADER_ELECTION_RENEW_DEADLINE" env-default:"10s"`
	LeaseDuration time.Duration `yaml:"leaseDuration" env:"LEADER_ELECTION_LEASE_DURATION" env-default:"15s"`
}

type MetricsConfig struct {
	Enabled *bool `yaml:"enabled" env:"METRICS_ENABLED" env-default:"false"`
	Port    int32 `yaml:"port" env:"METRICS_PORT" env-default:"8080"`
	Secure  *bool `yaml:"secure" env:"METRICS_SECURE" env-default:"false"`
}

type ProbeConfig struct {
	Enabled               *bool  `yaml:"enabled" env:"PROBE_ENABLED" env-default:"false"`
	Port                  int    `yaml:"port" env:"PROBE_PORT" env-default:"8082"`
	ReadinessEndpointName string `yaml:"readinessEndpointName" env:"PROBE_READINESS_ENDPOINT_NAME" env-default:"readyz"`
	LivenessEndpointName  string `yaml:"livenessEndpointName" env:"PROBE_LIVENESS_ENDPOINT_NAME" env-default:"healthz"`
}

type PprofConfig struct {
	Enabled *bool `yaml:"enabled" env:"PPROF_ENABLED" env-default:"false"`
	Port    int   `yaml:"port" env:"PPROF_PORT" env-default:"8081"`
}

type GlobalControllerConfig struct {
	SkipNameValidation      *bool          `yaml:"skipNameValidation" env:"GLOBAL_CONTROLLER_SKIP_NAME_VALIDATION"`
	GroupKindConcurrency    map[string]int `yaml:"groupKindConcurrency" env:"GLOBAL_CONTROLLER_GROUP_KIND_CONCURRENCY"`
	MaxConcurrentReconciles int            `yaml:"maxConcurrentReconciles" env:"GLOBAL_CONTROLLER_MAX_CONCURRENT_RECONCILES" env-default:"1" validate:"min=1"`
	CacheSyncTimeout        time.Duration  `yaml:"cacheSyncTimeout" env:"GLOBAL_CONTROLLER_CACHE_SYNC_TIMEOUT" env-default:"2m"`
	NeedLeaderElection      *bool          `yaml:"needLeaderElection" env:"GLOBAL_CONTROLLER_NEED_LEADER_ELECTION"`
}

type SpecificControllersConfig struct {
	Casdoor CasdoorControllerConfig `yaml:"casdoor"`
}

type CasdoorControllerConfig struct {
	SkipNameValidation      *bool         `yaml:"skipNameValidation" env:"CASDOOR_CONTROLLER_SKIP_NAME_VALIDATION"`
	MaxConcurrentReconciles int           `yaml:"maxConcurrentReconciles" env:"CASDOOR_CONTROLLER_MAX_CONCURRENT_RECONCILES" env-default:"1" validate:"min=1"`
	CacheSyncTimeout        time.Duration `yaml:"cacheSyncTimeout" env:"CASDOOR_CONTROLLER_CACHE_SYNC_TIMEOUT" env-default:"2m"`
	NeedLeaderElection      *bool         `yaml:"needLeaderElection" env:"CASDOOR_CONTROLLER_NEED_LEADER_ELECTION"`
}
